client/my-sites/domains/domain-management/list/bulk-edit-contact-info.jsx

```diff
@@ -0,0 +1,218 @@
+import FormLabel from 'calypso/components/forms/form-label';
+import FormCheckbox from 'calypso/components/forms/form-checkbox';
+import DesignatedAgentNotice from 'calypso/my-sites/domains/domain-management/components/designated-agent-notice';
+import { isFetchingUserSettings } from 'calypso/state/user-settings/selectors';
+
+import wp from 'calypso/lib/wp';
+
+const wpcom = wp.undocumented();
+
+/**
+ * Style dependencies
+ */
+import './list-all.scss';
+
+class BulkEditContactInfo extends React.Component {
+	static propTypes = {
+		handleSaveContactInfo: PropTypes.func.isRequired,
+		isDisabled: PropTypes.bool,
+		isSubmitting: PropTypes.bool,
+		onTransferLockOptOutChange: PropTypes.func.isRequired,
+		emailOnly: PropTypes.bool,
+	};
+
+	static defaultProps = {
+		isDisabled: false,
+		isSubmitting: false,
+		emailOnly: false,
+	};
+
+	state = {
+		hasSetContactDetailsFromCache: false,
+		contactDetails: {},
+		errorMessages: {},
+	};
+
+	componentDidMount() {
+		if ( isEmpty( this.props.userSettings ) ) {
+			this.props.fetchUserSettings();
+		}
+
+		if ( isEmpty( this.props.contactDetailsCache ) ) {
+			this.props.requestContactDetailsCache();
+		}
+
+		this.debouncedValidateContactDetails = debounce( this.validateContactDetails, 500 );
+	}
+
+	componentDidUpdate() {
+		const email = ! isEmpty( this.props.userSettings?.new_user_email )
+			? this.props.userSettings?.new_user_email
+			: this.props.userSettings?.user_email;
+
+		if ( ! isEmpty( email ) && isEmpty( this.state.contactDetails?.email ) ) {
+			this.updateDomainContactFields( { email: email } );
+		}
+
+		if (
+			! isEmpty( this.props.contactDetailsCache ) &&
+			false === this.state.hasSetContactDetailsFromCache
+		) {
+			this.setContactDetailsFromCache( this.props.contactDetailsCache );
+		}
+	}
+
+	isLoading() {
+		return (
+			isEmpty( this.props.contactDetailsCache ) ||
+			this.props.isRequestingContactDetailsCache ||
+			this.props.isFetchingUserSettings
+		);
+	}
+
+	setContactDetailsFromCache = ( data ) => {
+		delete data.email;
+		this.setState( { hasSetContactDetailsFromCache: true }, () => {
+			this.updateDomainContactFields( data );
+		} );
+	};
+
+	updateDomainContactFields = ( data ) => {
+		const newContactDetails = merge( {}, this.state.contactDetails, data );
+		this.setState( { contactDetails: newContactDetails } );
+		! isEmpty( newContactDetails ) && this.debouncedValidateContactDetails( newContactDetails );
+	};
+
+	onTransferLockOptOutChange = ( event ) => {
+		this.props.onTransferLockOptOutChange( event.target.checked );
+	};
+
+	validateContactDetails = ( contactDetails ) => {
+		if ( this.isLoading() ) {
+			return;
+		}
+
+		wpcom.validateDomainContactInformation( contactDetails, [], ( error, data ) => {
+			let errorMessages = ( data && data.messages ) || {};
+			if ( ! isEmpty( errorMessages ) ) {
+				errorMessages = mapValues( errorMessages, ( errors ) => {
```

Comment:
This is a little tricky if you want to preserve the original keys, like `mapValues` does. I wonder if this could be done later during the render process; when we'd render the errors, first check to see if there are multiple, and if so, join them then?

Otherwise, it is possible to map over the values and preserve the keys here. Something like:

```js
errorMessages = Object.keys( errorMessages ).reduce( ( messagesJoined, key ) => {
  const singleStringErrors = Array.isArray( errorMessages[ key ] ) ? errorMessages[ key ].join( ' ' ) : errorMessages[ key ];
  return {
    ...messagesJoined,
    [ key ]: singleStringErrors;
  };
}, {} );
```

Reply:
Here's what I ended up with...almost the same:
```
errorMessages = Object.entries( errorMessages ).reduce( ( result, [ field, errors ] ) => {
	result[ field ] = Array.isArray( errors ) ? errors.join( ' ' ) : errors;
	return result;
}, {} );
```